File: client/python/ws-subscribe.py
#!/usr/bin/env python3
"""
WebSocket Subscriber for sensor data
Requires: pip install websockets
"""

import json
import asyncio
from datetime import datetime
import websockets


async def subscribe():
    """Subscribe to sensor updates via WebSocket"""
    uri = "ws://localhost:9926/Sensors/101"

    try:
        async with websockets.connect(uri) as websocket:
            print("Connected to Harper WebSocket")
            print("Subscribed to: Sensors/101")
            print("Listening for messages...\n")

            # Receives the current state immediately upon connection,
            # then real-time updates as they're published
            async for message in websocket:
                try:
                    payload = json.loads(message)
                    # Harper sends change events with metadata, actual data is in 'value'
                    record = payload.get('value', payload)
                    timestamp = datetime.now().isoformat()

                    print(f"[{timestamp}] Update on Sensors/101:")
                    print(f"  Temperature: {record['temp']}°F")
                    print(f"  Location: {record['location']}")
                    print()

                except json.JSONDecodeError:
                    # If message isn't JSON, just log it as-is
                    timestamp = datetime.now().isoformat()
                    print(f"[{timestamp}] Update on Sensors/101: {message}")
                    print()
                except Exception as e:
                    print(f"Error processing message: {e}")
                    print()

    except KeyboardInterrupt:
        print("\nStopping subscriber...")
    except Exception as e:
        print(f"Error: {e}")


def main():
    asyncio.run(subscribe())


if __name__ == "__main__":
    main()

File: client/nodejs/ws-subscribe.js
const WebSocket = require('ws');

// Connect to Harper WebSocket endpoint for sensor 101
// To subscribe to all sensors, you would need to connect to each individually
// or use a custom Harper resource that broadcasts all sensor updates
const ws = new WebSocket('ws://localhost:9926/Sensors/101');

ws.on('open', () => {
  console.log('Connected to Harper WebSocket');
  console.log('Subscribed to: Sensors/101');
  console.log('Listening for messages...\n');
});

ws.on('message', (data) => {
  // Receives the current state immediately upon connection,
  // then real-time updates as they're published
  try {
    const payload = JSON.parse(data.toString());
    // Harper sends change events with metadata, actual data is in 'value'
    const record = payload.value || payload;
    console.log(`[${new Date().toISOString()}] Update on Sensors/101:`);
    console.log(`  Temperature: ${record.temp}°F`);
    console.log(`  Location: ${record.location}`);
    console.log('');
  } catch (err) {
    // If message isn't JSON, just log it as-is
    console.log(`[${new Date().toISOString()}] Update on Sensors/101:`, data.toString());
    console.log('');
  }
});

ws.on('error', (err) => {
  console.error('WebSocket error:', err);
  process.exit(1);
});

ws.on('close', () => {
  console.log('Connection closed');
  process.exit(0);
});

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\nStopping subscriber...');
  ws.close();
});
